Start to describe Movies API example (#6)
HydraCG/cookbook: movies/1.entry-point.md

# API entrypoint

## Story

As an API client I want to start with some URI to be able to discover the API capabilities.

## API behaviour

```
GET https://hydra-movies.herokuapp.com/
```

Response headers (excerpt):

```
Access-Control-Allow-Origin: *
Access-Control-Expose-Headers: Location, Link
Link: </doc>; rel="http://www.w3.org/ns/hydra/core#apiDocumentation"
```

Response body (excerpt):

```json
{
  "@context": "/context.jsonld",
  "@id": "/",
  "@type": "schema:EntryPoint"
}
```
### Explanation

> ```

Reviewer: ```suggestion
```http
```

GET https://hydra-movies.herokuapp.com/
```

The example API is hosted at https://hydra-movies.herokuapp.com/, so most commonly this is the URI we will give to clients initially. Clients that use the API for the first time will start by dereferencing this URI.

> ```

Reviewer: ```suggestion
```http
```

Access-Control-Allow-Origin: *
Access-Control-Expose-Headers: Location, Link
```

> As the API will be used by a variety of clients, we need to ensure a proper setup for Cross Origin Resource Sharing (CORS). The example API allows access from any Origin and exposes some relevant headers. Exposing the `Link` header is a **must have** for Hydra APIs, because the API documentation is discovered this way.

Reviewer: A link to https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS could be useful

Reviewer: > Exposing the `Link` header is a **must have**

Nit to the entire CORS thing - it only applies to browser clients. Maybe worth nothing that.

Author: rephrased the paragraph

> ```

Reviewer: ```suggestion
```http
```

Link: </doc>; rel="http://www.w3.org/ns/hydra/core#apiDocumentation"
```

> Hydra uses the HTTP `Link` header to refer to the API documentation. `</doc>` is a relative Link to the documentation and `rel="http://www.w3.org/ns/hydra/core#apiDocumentation"` is a fixed value that describes the type of the link. Every API resource (not only the entrypoint) must provide this header with every response. The client knows now, that the API documentation can be found at https://hydra-movies.herokuapp.com/doc.

Reviewer: Worth noting that failed requests should include the link too?

Author: yep

> Hydra uses the HTTP `Link` header to refer to the API documentation. `</doc>` is a relative Link to the documentation and `rel="http://www.w3.org/ns/hydra/core#apiDocumentation"` is a fixed value that describes the type of the link. Every API resource (not only the entrypoint) must provide this header with every response. The client knows now, that the API documentation can be found at https://hydra-movies.herokuapp.com/doc.

Reviewer: ```suggestion
Hydra uses the [HTTP `Link` header](http://webconcepts.info/concepts/http-header/Link) to refer to the API documentation. `</doc>` is a relative Link to the documentation and `rel="http://www.w3.org/ns/hydra/core#apiDocumentation"` is a fixed value that describes the type of the link. Every API resource (not only the entrypoint) must provide this header with every response. The client knows now, that the API documentation can be found at https://hydra-movies.herokuapp.com/doc.
```

```
"@context": "/context.jsonld"
```

> The example API uses JSON-LD as serialization format. This is not a necessity, as Hydra supports any RDF serialization, but it is an elegant way to extend tradtional JSON-APIs. The line above links to the JSON-LD context the API uses. The context will be described in more detail in the next recipe.

Reviewer: ```suggestion
The example API uses [JSON-LD](http://webconcepts.info/concepts/http-header/Link) as serialization format. This is not a necessity, as Hydra supports any RDF serialization, but it is an elegant way to extend tradtional JSON-APIs. The line above links to the JSON-LD context the API uses. The context will be described in more detail in the next recipe.
```

```
"@id": "/"
```

This is the ID of the entrypoint. All IDs in JSON-LD expand to full URIs, so this is equivalent to the URI the client dereferenced: https://hydra-movies.herokuapp.com/.

```
"@type": "schema:EntryPoint"
```

> Any resource (of any type) can be used as the entry point of a Hydra API. If your domain and API design allows it, you should use a more specific type. But sometimes the entry point has no further purpose on its own, and a [schema.org EntryPoint](https://schema.org/EntryPoint) can express this.

Reviewer: ```suggestion
Any resource (of any type) can be used as the entry point of a Hydra API. If your domain and API design requires it, you could use a more specific type. But sometimes the entry point has no further purpose on its own, and a [schema.org EntryPoint](https://schema.org/EntryPoint) can express this.
```

I propose to change the message here. I understand this more as  "**if needed, you can mint your own entrypoint type**"

Author: I would always recommend to use a specific type if it is possible I see schema:EntryPoint more as a fallback if nothing else fits. That`s why a formulated it that way.

Reviewer: Ok, point taken. Ignore this suggestion
